Issue1080 ROM soil temperature and interzonal heat exchange

PGorzalka opened the pull request:
Finally, I was able to finish the work on issue #1080 
@DaJansenGit & @FWuellhorst thanks for your support!

Changes mainly encompass the work mentioned in [this paper](https://doi.org/10.3384/ecp204577):
- building up on [changes already implemented via IBPSA merge](https://github.com/ibpsa/modelica-ibpsa/issues/1744), some adaptations for non-constant soil temperature
  - Option selector added to AixLib.BoundaryConditions
  - Options added to ZoneBaseRecord
  - Replaced TSoil constant in ThermalZone by option selector
- Interzonal heat exchange with new FiveElements RC model
  - Integration of that model into ThermalZone
  - Interzonal connection of zones in Multizone. For OM compatibility, I had to add an additional NzSplitter model and parameters to make sure arrays are initialized without errors.
  - Added ASHRAE test case 960

Additional minor stuff:
- made some parts of the ThermalZone model replaceable, so boundary conditions can be adapted more easily
- two missing `each`s in `AixLib.Fluid.Pools.BaseClasses.AirFlowMoistureToROM` caused OM to fail -> fixed

Missing:
- Test Case 960 was not actually tested against standard results because the weather file provided  (AixLib\Resources\weatherdata\ASHRAE140.mos) does not contain the weather data provided by the standard. 

See also [PR 785 in TEASER](https://github.com/RWTH-EBC/TEASER/pull/785)

Looking forward to your feedback on the changes!

FWuellhorst commented:
@PGorzalka , nice, thanks for adding this nice feature to AixLib! We have an extensive CI which does a lot of groundwork of the review, but it only runs for internal branches. Thus, I added an internal branch based on your fork. Now, we see if naming guidelines and regression tests all run. Depending on the changes, you either can work in your fork and merge your fork into the internal branch, or we could add you as a developer so you can work directly in AixLib @DaJansenGit ?

DaJansenGit commented:
@PGorzalka thanks for contributing!
@FWuellhorst Lets see what the CI outputs. If there are bigger changes that need to be addressed, I would add Philip as developer.

FWuellhorst commented:
@PGorzalka : You can find check and naming errors here:
https://rwth-ebc.github.io/AixLib/PGorzalka_issue1080_ROM-soil-and-zone-borders/index.html
With regard to the naming violations, please just address new code lines/files you added.

PGorzalka commented:
@FWuellhorst : Thank you for sharing the files.
I hope I solved all the namespace violations. The failed models are running now, too. However, the `AixLib.Controls.HVACAgentBasedControl.Examples.BuildingHeatingSystems.BuildingHeating` fails after 5 seconds. The same issue happens when I select the main branch, so I assume this is not due to my contributions?!

Should I add the example from TEASER as an additional test case? The ASHRAE Case 960 does not cover the ground temperature selection (yet, see question about weather data above).

FWuellhorst commented:
You can ignore the HVACAgent models.
However, the model check in ThermalZones.Examples causes Dymola to crash both in CI and locally on my PC. 
The last view before the crash: 
![image](https://github.com/user-attachments/assets/f1cec4c4-1d4e-4451-9f66-85d13a1cd7e0)

Aside from that, I added a default to FlowArrayRearranging in order to avoid a model check error:
```
  parameter Integer conPaiArr[nCon, 2]={{1, 2} for i in 1:nCon} "Array of port numbers to connect";
```

PGorzalka commented:
> the model check in ThermalZones.Examples causes Dymola to crash both in CI and locally on my PC. The last view before the crash: ![image](https://private-user-images.githubusercontent.com/22442514/387622882-f1cec4c4-1d4e-4451-9f66-85d13a1cd7e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMjQ5MjYsIm5iZiI6MTczMjAyNDYyNiwicGF0aCI6Ii8yMjQ0MjUxNC8zODc2MjI4ODItZjFjZWM0YzQtMWQ0ZS00NDUxLTlmNjYtODVkMTNhMWNkN2UwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE5VDEzNTcwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzMDg1ODQzOTI0MmVhN2VjYTZmNTU3NGE4N2UxYWRlN2NmMDljMDQzZjFmOGU1OTYxNDM2ZDhmZTVkNGQyZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.D0ON35ZPqam9Jazy_qk2_S2C6yY3714rXf5IU1A359Q)

This was due to the ZoneBaseRecords that the examples were using not having the new parameters. I added them there now. I did **not** add default values to the new ZoneBaseRecord parameters because the other parameters don't have default values either, including those recently added. Please let me know if I should set default parameters anyway to improve backwards compatibility.

> Aside from that, I added a default to FlowArrayRearranging in order to avoid a model check error:
> 
> ```
>   parameter Integer conPaiArr[nCon, 2]={{1, 2} for i in 1:nCon} "Array of port numbers to connect";
> ```
Thank you :)

PGorzalka commented:
@FWuellhorst thanks for the merge into the CI branch. There were some namespace violations left, which should now be resolved.

FWuellhorst commented:
Nice, the CI now passes. @DaJansenGit do you want to do the review, as actually now the ROM?
Only comment from my side: You could add a new example and a simulate and plot script (described here: https://github.com/RWTH-EBC/AixLib/pull/1548#issuecomment-2482094202) so that your features are tested every time we change something. 
Anyway, thanks for the feature!

DaJansenGit commented:
Thank you both for cleaning this up so quickly! I will do the review now

DaJansenGit commented on AixLib/DataBase/ThermalZones/ZoneBaseRecord.mo:
The changes made will not be downwards compatible as existing models won't have the the new values (`nIze` etc.). From my site this is fine.  The other option would be to create a new base `ZoneBaseRecord5Elements` and extend from this one here. But this also would lead to conditional statements in the `ThermalZone` model and in teaser and we will need to maintain both templates (TEASER) and records (modelica) in the future. So I vote for non downward compatibility and a new release. What do you think @FWuellhorst ?

FWuellhorst replied:
Why can't we just set the default to 0 or false to disable the internal connections?

PGorzalka replied:
thanks for implementing this; now also implemented in this branch with renamed `DataSourceGroTem`

DaJansenGit commented on AixLib/ThermalZones/ReducedOrder/RC/BaseClasses/ExteriorWallContainer.mo:
I don't really understand the name of this component. Why is it called `ExteriorWallContainer`? It is supposed to be used for interzonal heat transfer. Also the description of the instance in `AixLib.ThermalZones.ReducedOrder.RC.FiveElements` says “RC element for interzonal elements”. Can you please explain this in more detail?

PGorzalka replied:
The RC elements in the current `FourElements` model, no matter if floor, roof, or exterior wall, are all `ExteriorWall` instances. The container is needed because the heat flow is passed through on one of the two zones that is connected - making sure that the same element is not modelled twice. I did not want to call the container explicitly `InterzonalElementContainer` because it could have another use. Would you rather call it like that?

DaJansenGit commented on AixLib/ThermalZones/ReducedOrder/RC/FiveElements.mo:
For some reason, all the new graphical representations of the elements you added for the adjacent zone heat transfer implementation are placed in the center of the diagram (see screenshot)
![image](https://github.com/user-attachments/assets/6283bce5-3148-407d-8998-1a6f65a30e76)

PGorzalka replied:
done. Seems that OM deleted parts of the annotations made by Dymola previously...

DaJansenGit reviewed (changes requested):
Somehow most or all the components you have added are placed in the center of the diagram in all models. I don't know what happened here, but for some models this worsens the visibility considerably and it would be good if you could move the models to a suitable position.

Regarding the implementation this looks fine. Only thing t.b.d. from my site is regarding the extension of the record. Let's see what @FWuellhorst thinks about this.

PGorzalka commented:
Default values would be feasible I think, I just did not implement it because there are none for Three/FourElement parameters either. I’d stick to whatever you suggest.
Am 26.11.2024 um 17:12 schrieb FWuellhorst ***@***.***>:

﻿

@FWuellhorst commented on this pull request.

________________________________

On AixLib/DataBase/ThermalZones/ZoneBaseRecord.mo<https://github.com/RWTH-EBC/AixLib/pull/1546#discussion_r1858834274>:

Why can't we just set the default to 0 or false to disable the internal connections?

—
Reply to this email directly, view it on GitHub<https://github.com/RWTH-EBC/AixLib/pull/1546#discussion_r1858834274>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AIQSKT23HXEQI62VYWEPVMT2CSMXXAVCNFSM6AAAAABR6BW3T6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINRSGAZDMOBQHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

FWuellhorst commented:
Ok, I never used something other than four elements. But, as this is the current practice, I would just add nIze=1 and the default values to the TEASER export, and specify it in the BaseRecord to be backwards compatible.
![image](https://github.com/user-attachments/assets/42dd9609-8d30-4352-8f22-609d4df5c8c1)

On the long run, I think the record would benefit by using groups or tabs to shows users the different elements, use conditions, etc. I always disliked the record as a sort of global parameter object with no structure, and new (cool) features like this further increase complexity. But we should tackle this in a separate issue @DaJansenGit, or? Should be fast, as it is purely graphical. We could maybe add Booleans to disable the parameter view and avoid the A > 0 cases in the ROM, but that would not be compatible and require IBPSA changes, so no :D

DaJansenGit commented:
> Ok, I never used something other than four elements. But, as this is the current practice, I would just add nIze=1 and the default values to the TEASER export, and specify it in the BaseRecord to be backwards compatible. ![image](https://private-user-images.githubusercontent.com/22442514/390346152-42dd9609-8d30-4352-8f22-609d4df5c8c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3NzU2NjAsIm5iZiI6MTczMjc3NTM2MCwicGF0aCI6Ii8yMjQ0MjUxNC8zOTAzNDYxNTItNDJkZDk2MDktOGQzMC00MzUyLThmMjItNjA5ZDRkZjVjOGMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI4VDA2MjkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2NDIwM2JkNjViZGRmMGY2MGJkZjJmZGVlNzVjZGNmOTBkZDI1MDk4OWYyYWYxYmM4Y2YzMGQxMzNiNDI2MDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nNMRts1LA3Be9dFg4HAzMP1EKGsRkVTt0VYFcmpelcs)
> 
> On the long run, I think the record would benefit by using groups or tabs to shows users the different elements, use conditions, etc. I always disliked the record as a sort of global parameter object with no structure, and new (cool) features like this further increase complexity. But we should tackle this in a separate issue @DaJansenGit, or? Should be fast, as it is purely graphical. We could maybe add Booleans to disable the parameter view and avoid the A > 0 cases in the ROM, but that would not be compatible and require IBPSA changes, so no :D

I agree with the proposed solution to adjust the TEASER export and BaseRecord, so its backwards compatible. I did the changes in the CI branch and if CI passes I will merge it into this branch. In TEASER @PGorzalka already did add the changes to the export, so we should be fine.
Furthermore, I really like the idea of adding some dialog section to increase readability of the record.

FWuellhorst commented:
@DaJansenGit @PGorzalka : I updated the CI in PR https://github.com/RWTH-EBC/AixLib/pull/1554 to Dymola2024x. I would wait to merge this depending on your timeline. Maybe some regression results would need updating if they even exist yet. If they don't  exist yet, this PR should not affect this PR.

PGorzalka commented on AixLib/BoundaryConditions/GroundTemperature/DataSource.mo:
done

PGorzalka commented on AixLib/BoundaryConditions/GroundTemperature/Options.mo:
seems to work now, could you please check again?

PGorzalka commented:
@FWuellhorst @DaJansenGit I added the example and the respective script. I hope I did it correctly. But I honestly don't know how to handle the reference results. The explanation [here](https://github.com/RWTH-EBC/AixLib/wiki/Test-Management) seems to be outdated and BuildingsPy no longer used - at least the mentioned runUnitTests.py script is no longer presented in the AixLib repository?!

FWuellhorst commented:
@PGorzalka : mhm, this seems to be outdated indeed. Our CI pipeline generates the results for you. 
This PR comment explains the new process: https://github.com/RWTH-EBC/AixLib/pull/1548#issuecomment-2482094202 but is not in this PR as the CI does not run for forks. Lets see if the scripts work!

FWuellhorst commented:
The regression plots are displayed here: https://rwth-ebc.github.io/AixLib/PGorzalka_issue1080_ROM-soil-and-zone-borders/charts/ThermalZones/AixLib_ThermalZones_ReducedOrder_Examples_MultizoneInterzonalsFixedHeater.html
Are the fluctuations in zone 1 intended? I had to change to cvode solver, as radau gets timeouts in CI.

Some other results seem to have changed translation statistics. Without interzonal exchange, the change in soil temperature model could cause this, or?

PGorzalka commented:
Thank you for checking the results!

Regarding fluctuations: this is indeed intended. Zone 1 is an unheated attics with high infiltration.

Regarding translation: yes, this could be caused by the soil temperature model. Additionally, the FiveElement model has replaced the FourElement in the Multizone.

FWuellhorst commented:
Ok, it seems to only affect this model:
```
*** Warning: AixLib.ThermalZones.ReducedOrder.Examples.MultizoneMoistAirCO2: Translation statistics for simulation changed for linear, but results are unchanged.
 Old = 4, 0, 4, 0, 4, 0, 4, 0, 4, 0
 New = 3, 0, 3, 0, 3, 0, 3, 0, 3, 0
```
I will update these statistics, as smaller is better. Then you can update your fork with the main and the aixlib-internal fork branch.

FWuellhorst commented:
@PGorzalka it seems that your model takes more than 300 s to solve in CI, where multiple models run in parallel. Running it locally in docker on a single core, it works fine. 4 processors work fine as well, but I guess this is the problem, because various models get timeouts which worked fine before. 
So, my idea would be to decrease the simulation time to maybe 3-4 days? As long as it is a phase with interzonal exchange, it should be sufficient for regression test purposes.

PGorzalka commented:
@FWuellhorst i changed the T_start in the example so it requires less time to initialize. Simulated time is now 140 hours to have at least a few of the daily dynamics included. I think I removed the reference results, but I'm not sure. Can you have a look if it works now?

FWuellhorst commented:
@PGorzalka Thanks. Now, the CI runs. @DaJansenGit From my side the changes look good, do you have open issues? 
@PGorzalka , you have to merge the new reference results into your fork if we merge your fork.

PGorzalka commented:
@FWuellhorst @DaJansenGit thanks! Did that. I also added our paper to the publications list.

DaJansenGit reviewed (approved):
Everything looking fine now. I just updated the publications.md in your branch as it was faulty formatted. Thanks again for the extensive change to the model.

DaJansenGit commented:
I just tested some examples again before we do the final merge and found these warnings (screenshot) occurring, when simulating an previously existing example like the `AixLib.ThermalZones.ReducedOrder.Examples.Multizone` that uses existing ThermalZone records (in this case `AixLib.DataBase.ThermalZones.OfficePassiveHouse.OPH_1_Office`) 

![image](https://github.com/user-attachments/assets/137ceb31-e5de-4a3c-beb6-c25c69d67bd3)

I have to admit that I do not understand why the warning with the size of `AIze` pops up, because you gave NIze and AIze legit default values in the base record...

PGorzalka commented:
It's because the size of `AIze` is defined by `nIze` and the latter cannot be resolved before the size check for some reason. So it fails to compare the required size `1` to the not-yet resolved `nIze`. Fortunately it throws just a warning :)

Imho, this should be fine. @FWuellhorst what do you think?

FWuellhorst commented:
I think the issue is due to the vector approach of zoneParam in Multizone. Records are sometimes checked last, and the values of `AIze` are used as structural paramters with `ATotIze`. Removing the `final` in `final nIze=zoneParam.nIze` and then overwriting `each nIze=1` in the examples removes the warning. So, if we want to get rid of this warning at some point, nIze would need to be outside the record, as well. But for now let's keep the warnings, as they don't lead to errors.

PGorzalka commented:
So there is nothing left to do, am I right @DaJansenGit @FWuellhorst ?

FWuellhorst commented:
@PGorzalka can you merge branch `PGorzalka_issue1080_ROM-soil-and-zone-borders` into yours? I updated it with the latest main. Then you can merge :)

PGorzalka commented:
> @PGorzalka can you merge branch `PGorzalka_issue1080_ROM-soil-and-zone-borders` into yours? I updated it with the latest main. Then you can merge :)

 @FWuellhorst done :) but I can't merge, you have to
